easydock/database.py

```diff
@@ -405,6 +405,166 @@ def select_mols_to_dock(db_conn, table_name='mols', add_sql=None):
+    '''
+    Protonate SMILES by Chemaxon cxcalc utility to get molecule ionization states at pH 7.4
+    :param db_fname:
+    :param program: name of the prorgam to use
+    :param tautomerize: get a major tautomer at protonation
+    :param table_name: table name with molecules to protonate
+    :param add_sql: additional SQL query to be appended to the SQL query to retrieve molecules for protonation,
+                    e.g. "AND id IN ('MOL1', 'MOL2')" or "AND iteration=(SELECT MAX(iteration) FROM mols)".
+    :return:
+    '''
+    conn = sqlite3.connect(db_fname, check_same_thread=False)   # danger
+
+    try:
+        cur = conn.cursor()
+
+        # process SMILES and mol_block together
+        sql = f"""SELECT smi, id || '_' || stereo_id 
+                  FROM {table_name} 
+                  WHERE smi IS NOT NULL AND docking_score is NULL AND smi_protonated is NULL"""
+        sql += add_sql
+        data_list = list(cur.execute(sql))
+
+        if not data_list:
+            sys.stderr.write(f'no molecules to protonate\n')
+            return
+
+
+        cur.execute(sql)
+        if program in ['chemaxon']:  # file-based protocol, files are created by chunks
+            if program == 'chemaxon':
+                protonate_func = partial(protonate_chemaxon, tautomerize=tautomerize)
+                read_func = read_protonate_chemaxon
+            else:
+                raise ValueError(f'There is no implemeneted functions to protonate molecules by {program}')
+
+            nmols = ncpu * 500  # batch size
+            while True:
+                with tempfile.NamedTemporaryFile(suffix='.smi', mode='w', encoding='utf-8') as tmp:
+                    res = cur.fetchmany(nmols)
+                    if not res:
+                        break
+                    else:
+                        for smi, mol_name in res:
+                            tmp.write(f'{smi}\t{mol_name}\n')
+                    tmp.flush()
+
+                    fd, output = tempfile.mkstemp()
+                    try:
+                        protonate_func(tmp.name, output)
+                        items = read_func(output)  #  generator of tuples (smi, mol_name)
+                        update_db_protonated_smiles(conn, items, data_list, table_name)
+                    finally:
+                        os.remove(output)
+                        os.close(fd)
+
+        elif program in ['pkasolver']:  # native python protocol
+            if program == 'pkasolver':
+                protonate_func = partial(protonate_pkasolver, ncpu=ncpu, smi_size=len(data_list))
+            else:
+                raise ValueError(f'There is no implemeneted functions to protonate molecules by {program}')
+
+            items = []
+            for i, item in enumerate(protonate_func(cur), 1):  # cursor is a generator like object, it can be used as input instead of read_input(input_fname)
```

Comment:
Replace with these two lines:
```python
data = ((smi, mol_name) for smi, mol_name in cur)
for i, item in enumerate(protonate_func(data), 1):
```
This will make data structure more explicit and should not cause an overhead due to lazy evaluation of a created generator